### PR TITLE
Propose rules for team repo maintenance

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Per leadership council discussion in our last meeting, proposing a set of rules for us to approve on behalf of the project with regards to team repo administration. The goal is to largely codify the status quo, while providing opportunity for expansion of merge rights under documented guidelines.

cc @jackh726

### CODE REVIEW

File: src/infra/team-maintenance.md [hunk 1]
```diff
@@ -4,34 +4,64 @@ The roster of the Rust teams is always in flux. From time to time, new people
...
+### Who belongs to team-admins?
+
+This group of people is nominated & approved by the Leadership Council, but is
+not selected through any formal criteria. Eventually, we hope that the need for
+this group to exist will diminish as additional automation is added to enforce
+the above policies.
+
+Note also that the [infra-admins] team maintains "root" credentials to Rust
+infrastructure, including the team repo, in order to make changes if needed to
+keep infrastructure opertaional. Those rights should only be exercised when
```
Comment (rylev) [inline src/infra/team-maintenance.md:54]:
```suggestion
keep infrastructure operational. Those rights should only be exercised when
```

File: src/infra/team-maintenance.md [hunk 2]
```diff
@@ -4,34 +4,64 @@ The roster of the Rust teams is always in flux. From time to time, new people
...
+    which team owns the repository, when in doubt asking for clarification and
+    codifying in a comment in the relevant repository.
+
+### Source code changes
+
+The team repository additionally contains code to transform and validate the
+TOML user-edited files. This is owned by the Infrastructure team and approval
+should be sought for changes.
+
+### Who belongs to team-admins?
```
Comment (rylev) [inline src/infra/team-maintenance.md:45]: Nit: `team-admins` might not be the best name, but I don't have a great alternative. Perhaps the more verbose `team-repo-admins`?

File: src/infra/team-maintenance.md [hunk 3]
```diff
@@ -40,14 +70,12 @@ To make a full team member, the following places need to be modified:
...
-- [toolstate notifications]
 - `triagebot.toml` files of all repos they were involved in
+- 1password
 
 [gh-team]: https://github.com/orgs/rust-lang/teams
 [gh-nursery-team]: https://github.com/orgs/rust-lang-nursery/teams
 [team repo]: https://github.com/rust-lang/team/tree/master/teams
 [team website]: https://www.rust-lang.org/governance
-[toolstate notifications]: https://github.com/rust-lang/rust/blob/master/src/tools/publish_toolstate.py
+[team-admins]: TODO
```
Comment (rylev) [inline src/infra/team-maintenance.md:81]: Is the plan to open a PR against the team repo and then link once that's merged?

Comment (Mark-Simulacrum) [reply src/infra/team-maintenance.md:81]: Yes, that's my goal. I think we can discuss exact membership at that point - I don't have any strong feelings on that.

**Comment by Mark-Simulacrum** [comment]
@rfcbot merge

Kicking off FCP on this. There's some comments on details (naming and such) but I think we can gather feedback in the meantime.

File: src/infra/team-maintenance.md [hunk 4]
```diff
@@ -4,34 +4,64 @@ The roster of the Rust teams is always in flux. From time to time, new people
...
-- Mailgun email lists
+# Rules for changes to team repo
 
-Team membership is duplicated in a few other places listed below, but the
-long-term goal is to centralize on the team repo.
+Pull requests to the repository are merged by the [team-admins], who use these rules to merge PRs:
+
+### `people`, `teams`, and `repos` directories
+
+If a change is related to an individual and does not expand permissions, then only the individual's approval is required. If the change has already been made outside of the team repo (e.g., GitHub username change) then it is considered implicitly approved. This non-exhaustively includes:
```
Comment (rylev) [inline src/infra/team-maintenance.md:21]: > and does not expand permissions

I'm not sure it's always easy to tell when permissions are impacted or not. Like some marker teams are just markers for notifications but others have permissions associated with them. Are we running the risk of someone thinking a change doesn't impact permissions but in reality it does?

Comment (Mark-Simulacrum) [reply src/infra/team-maintenance.md:21]: Ideally, everything ends up being tied in to the team repo long-term so this becomes a CI check or similar... but I'm not sure we can get there quickly.

My sense is that the answer probably should be organizational for now - i.e., don't tie permissions to teams that aren't obviously permission-d. When in doubt checking with infra and/or escalating to a lead seems ok.

Do you feel like this comes up in practice? I feel like I'm usually pretty confident in what the implications of a change to team repo are. Maybe that's false confidence though :)

Comment (rylev) [reply src/infra/team-maintenance.md:21]: I similarly feel pretty confident, but I echo your desire to get to just a CI check in the long term and this is just me worrying about the time between now and then.

Comment (Mark-Simulacrum) [reply src/infra/team-maintenance.md:21]: I'm going to go ahead and mark this thread resolved. I think we want the CI check anyway (no need for that to be coupled here) and it should be OK in practice in the meantime. Folks can always fallback on the more expensive "get team lead to approve" strategy.

File: src/infra/team-maintenance.md (hunk 2 above)
Comment (Mark-Simulacrum) [inline src/infra/team-maintenance.md:45]:
```suggestion
### Who belongs to team-repo-admins?
```

File: src/infra/team-maintenance.md [hunk 5]
```diff
@@ -4,34 +4,64 @@ The roster of the Rust teams is always in flux. From time to time, new people
...
+
+This group of people is nominated & approved by the Leadership Council, but is
+not selected through any formal criteria. Eventually, we hope that the need for
+this group to exist will diminish as additional automation is added to enforce
+the above policies.
+
+Note also that the [infra-admins] team maintains "root" credentials to Rust
+infrastructure, including the team repo, in order to make changes if needed to
+keep infrastructure operational. Those rights should only be exercised when
+required though, with team-admins being the first point of contact for changes.
```
Comment (Mark-Simulacrum) [inline src/infra/team-maintenance.md:55]:
```suggestion
required though, with team-repo-admins being the first point of contact for changes.
```

File: src/infra/team-maintenance.md [hunk 6]
```diff
@@ -4,34 +4,64 @@ The roster of the Rust teams is always in flux. From time to time, new people
...
 
-- the [team website]
-- bors r+ rights
-- rfcbot interaction
-- Mailgun email lists
+# Rules for changes to team repo
 
-Team membership is duplicated in a few other places listed below, but the
-long-term goal is to centralize on the team repo.
+Pull requests to the repository are merged by the [team-admins], who use these rules to merge PRs:
```
Comment (Mark-Simulacrum) [inline src/infra/team-maintenance.md:17]:
```suggestion
Pull requests to the repository are merged by the [team-repo-admins], who use these rules to merge PRs:
```

File: src/infra/team-maintenance.md [hunk 7]
```diff
@@ -4,34 +4,64 @@ The roster of the Rust teams is always in flux. From time to time, new people
...
+
+If a change will grant additional permissions, then a team lead needs to
+approve the change. Any team lead in the "parent team" chain may do so. This includes:
+
+* Adding new subteams under an existing team
+* Granting additional access (bors, dev-desktop access, etc.)
+* Changing other metadata (website descriptions, Zulip groups, etc.)
+* Changing access to repositories owned by their team
+  * For repositories ownership is not currently formally tracked. Until that is
+    added, the team-admins are expected to exercise their understanding of
```
Comment (Mark-Simulacrum) [inline src/infra/team-maintenance.md:35]:
```suggestion
    added, the team-repo-admins are expected to exercise their understanding of
```

File: src/infra/team-maintenance.md (hunk 3 above)
Comment (Mark-Simulacrum) [inline src/infra/team-maintenance.md:81]:
```suggestion
[team-repo-admins]: TODO
```

Comment (Mark-Simulacrum) [inline src/infra/team-maintenance.md]:
```suggestion
[team-admins]: https://github.com/rust-lang/team/blob/master/teams/team-repo-admins.toml
```

File: src/infra/team-maintenance.md [hunk 8]
```diff
@@ -40,14 +70,12 @@ To make a full team member, the following places need to be modified:
...
-- [toolstate notifications]
 - `triagebot.toml` files of all repos they were involved in
+- 1password
 
 [gh-team]: https://github.com/orgs/rust-lang/teams
 [gh-nursery-team]: https://github.com/orgs/rust-lang-nursery/teams
 [team repo]: https://github.com/rust-lang/team/tree/master/teams
 [team website]: https://www.rust-lang.org/governance
-[toolstate notifications]: https://github.com/rust-lang/rust/blob/master/src/tools/publish_toolstate.py
+[team-repo-admins]: TODO
```
Comment (jackh726) [inline src/infra/team-maintenance.md:81]:
```suggestion
[team-repo-admins]: https://github.com/rust-lang/team/blob/master/teams/team-repo-admins.toml
[infra-admins]: https://github.com/rust-lang/team/blob/master/teams/infra-admins.toml
```

**Review by jackh726 — APPROVED** [review]
Seems like these links got removed, maybe in rebase?

**Pull request merged**